### PR TITLE
fix: remove every row scoped to an event when the event is deleted

### LINKED ISSUE #458: admin: deleting an event leaves orphaned rows in eight tables
## Describe the Bug

Deleting an event from the admin panel leaves rows in seven tables still pointing at the deleted event, room, booth and segment ids.
The panel reports success.

`PRAGMA foreign_keys` is 0 on this SQLite setup, so the `ondelete="CASCADE"` declarations in `portal/models.py` never fire.
Only children reachable through an ORM `cascade="all, delete-orphan"` relationship are actually removed, which covers `rooms`, `booths`, `invite_tokens`, `booth_memberships` and the booth/room translation languages.
Everything else stays.

`delete_room` (`portal/database.py:235`) already works around this with explicit deletes, and a comment naming the reason:

```python
# First delete booth memberships and tokens to avoid orphans since SQLite FKs are OFF
```

`delete_event` (`portal/database.py:173`) has no equivalent.

## Steps to Reproduce

1. Create an event with a room and a booth.
2. Give it an event membership, a room membership, a transcript segment with a translation, and an OAuth token, consent grant and authorization code scoped to that event.
3. Go to Admin -> Events and delete the event.
4. Query the tables below.

Measured on a fresh in-memory DB, deleting through `POST /admin/events/1/delete`, which returned 303:

```
PRAGMA foreign_keys = 0

table                        before  after   status
events                            1      0   gone
rooms                             1      0   gone
booths                            1      0   gone
invite_tokens                     1      0   gone
booth_memberships                 1      0   gone
transcript_segments               1      1   ORPHANED
transcript_translations           1      1   ORPHANED
event_memberships                 1      1   ORPHANED
room_memberships                  1      1   ORPHANED
oauth_tokens                      1      1   ORPHANED
oauth_consent_grants              1      1   ORPHANED
oauth_authorization_codes         1      1   ORPHANED
oauth_audit_logs                  1      1   ORPHANED
```

**Correction, 2026-09-07.** The title and the table above say seven tables; it is eight.
I built the original fixture by hand and it did not cover `oauth_audit_logs`.
Re-deriving the set from `Base.metadata` by transitive foreign-key closure gives 16 event-scoped tables, and `oauth_audit_logs` is the one I had missed.

It also needs the opposite treatment.
All three of its foreign keys are `ondelete="SET NULL"` and nullable, so the schema asks for the audit row to outlive what it refers to.
The right fix is to null `event_id` and `token_id` and keep the row, not delete it.
`client_id` is not event scoped and should be left alone.

Two further cases the original report missed, both found while fixing this:

- `transcript_segments` reaches the event through `booth_id` as well as `room_id`, and the two can disagree, because `admin_create_booth` takes both ids from the URL without checking that the room belongs to the event. Scoping cleanup by `room_id` alone leaves a segment holding a freed booth id, and the next booth created takes that id.
- `delete_room` strands `transcript_segments`, `transcript_translations`, `room_memberships` and `booth_translation_languages`. `delete_event` only reaches those through the event's rooms and booths, so anything stranded by a room delete can never be cleaned up afterwards.

## Expected Behavior

Deleting an event removes every row scoped to it, the way `delete_room` does for the subtree it owns.

## Additional Context

Found while fixing #456.
The leak is identical with and without a relay booth set, so it predates that fix.
One interaction is worth noting though: events that previously could not be deleted at all can now be deleted, and they leak the same way.

Two plausible directions, and I do not know which the project prefers:

1. Turn on `PRAGMA foreign_keys` for SQLite connections and let the existing `ondelete="CASCADE"` declarations do the work they were written for.
   Broadest fix, but it enables enforcement across every existing delete path at once, so it needs its own testing pass.
2. Add ORM cascades or explicit deletes for the seven tables, matching what `delete_room` already does.

I measured direction 1 rather than guessing at it.
Forced on for every engine the test suite creates, `pytest tests/` gives 4 failed and 68 errors.
The 68 are `Base.metadata.drop_all` hitting the pre-existing `booths` <-> `rooms` cycle that SQLAlchemy already warns about, and the 4 are webhook fixtures inserting rows with a dangling `developer_account_id`.
So it is worth doing but needs its own pass, exactly as suspected above.

#463 takes direction 2. Say the word if you would rather have direction 1 and I will send that instead.

### PR DESCRIPTION
**Pull request by arpan7sarkar**
## Description

**Update after rebasing on main (read this first).**
#457, which this PR was stacked on, has been squash merged as `0204679`, so the diff below is now only this PR's own four commits.
#477 has also merged, and it turns on `PRAGMA foreign_keys=ON` for every engine the app creates.
The original premise of this PR was that SQLite never enforces the `ondelete=` rules, and that is no longer true on main.
I re-measured everything against current main instead of reusing the old numbers, and the honest result is:

- **The orphan cleanup is no longer needed by the running app.**
  On current main, deleting an event through the admin panel already leaves 0 orphaned rows, because the database now cascades the deletes itself.
  The explicit statements in this PR still run and still pass, but on the app's own engine they now duplicate what SQLite does.
- **The permission fix is still needed.**
  On main, a `room_coordinator` of one side room can still delete the whole event, and with #477 that now destroys every transcript, membership and OAuth grant for it.
  This PR returns 403 for that and touches nothing.

So this PR is now mainly the guard change, with the orphan cleanup as defense in depth.
I am happy to go either way: keep it as is, or cut it down to the guard change and its two tests and drop the `delete_event` / `delete_room` changes.
Just say which.

## Related Issue

Closes #458.
The reachable part of #458 is already fixed on main by #477.
This PR adds explicit cleanup on top of that, so it does not depend on the pragma being set.

## Changes Made

**1. `POST /admin/events/{id}/delete` requires the event owner** (`portal/routers/admin.py`).
It used `require_admin`, which admits a `room_coordinator` of any single room as an admin for the whole event.
It now uses `require_event_owner`, which the API key routes on the same resource already use.
That still admits the `admin_token`, a site admin and the event's own owner.
Only the coordinator case changes.

**2. `delete_event` and `delete_room` clear their scoped rows explicitly** (`portal/database.py`).
When this was written, `PRAGMA foreign_keys` was 0, so only children reachable through an ORM cascade were removed and eight tables kept rows pointing at freed ids.
With #477 the database handles that on the app engine, so this is now a second layer rather than the fix.
It does still cover any engine created without the pragma.
Right now that means only the test fixtures, since `_get_engine()` and `configure()` both set it.
Three of the cases need more than a plain delete by `event_id`:

- **`oauth_audit_logs` is updated, not deleted.**
  All three of its foreign keys are `SET NULL` and nullable, so the audit row is meant to outlive what it refers to.
  `event_id` and `token_id` are nulled in two separate statements, because one statement with an `OR` would damage a second event's rows.
  `client_id` is not event scoped and is left alone.
- **`transcript_segments` is scoped by `room_id` OR `booth_id`.**
  It reaches the event through both, and they can disagree because `admin_create_booth` takes both ids from the URL without checking that the room belongs to the event.
- **`rooms.relay_booth_id` is scoped by booth, not by the room's event**, for the same unvalidated form reason via `admin_edit_room`.
  The relationship is assigned rather than the column, so a `Room` already in the session drops the edge too.

One effect of this code that the pragma does not give you: it also clears these pointers on objects already loaded in the current session.
With only the database `SET NULL`, a `Room` or `OAuthToken` loaded earlier in the same session still shows the old id until it is refreshed.
Every request gets a fresh session, so this does not reach users today.

The code comments in `delete_event` and `delete_room` still say SQLite FKs are OFF.
That is now out of date, and I will reword or remove them depending on which way you want to take this PR.

## Testing

End to end on a real uvicorn server, with an alembic-migrated SQLite database, a real admin login through `/admin/login` and a real coordinator login through `/login`.
The database was seeded with two identical events, so over-deletion would show.
The event being deleted has its own relay booth, and the surviving event has a room relaying one of its booths.

| | current `main` (with #477) | this PR |
|---|---|---|
| `room_coordinator` of one side room deletes the event | **303, whole event deleted** | **403, nothing touched** |
| admin deletes the event | 303 | 303 |
| rows left scoped to the deleted event | 0 | 0 |
| orphaned rows anywhere | 0 | 0 |
| `PRAGMA foreign_key_check` | empty | empty |
| surviving event's rows | all intact | all intact |
| surviving room's cross-event relay pointer | nulled | nulled |

I also ran this PR's new tests against main's code:

- All 10 fail on main with the tests exactly as written. 9 are in `tests/test_database.py`, whose `db` fixture builds its own engine without the pragma, so they test the explicit cleanup rather than what production does.
  The 10th is `test_delete_event_rejects_a_room_coordinator`, which fails for the real reason.
- When that fixture turns the pragma on the way `portal/database.py` does, 6 of the 9 pass on main.
  The other 3 fail only because an object loaded earlier in the same session still shows the old id, as described above.
  That run also produces 71 teardown errors, because `Base.metadata.drop_all` hits the existing `booths` <-> `rooms` cycle once foreign keys are on.
  The `db` fixture no longer matches production and would need that fixed first.

The screenshot below is from the original run, before #477 merged.

image

- [x] Tested locally, end to end as above
- [x] Existing tests pass: 572 passed on this branch after the rebase
- [x] Added/updated tests: 8 in `tests/test_database.py` and 2 in `tests/test_admin_panel.py`
- [x] CI green after the rebase

## Checklist

- [x] My code follows the project's coding standards
- [x] I have tested my changes
- [ ] I have updated documentation where necessary - no docs describe this behaviour
- [x] My changes do not introduce new warnings or errors - `ruff check` clean.
  `ruff format --check` now reports 25 files on main, all from other merged PRs (including the pragma blocks #477 added to `portal/database.py`); none of them are lines this PR changes.

## Additional Notes

- **An event soft-deleted through the v1 API can never be hard-deleted.**
  `DELETE /api/v1/events/{slug}` sets `deleted_at`, and `get_event_by_id` filters `deleted_at IS NULL`, so `delete_event` returns `False` while `admin_delete_event` discards the return and redirects as if it succeeded.
  This is pre-existing and still true on main; #460 tracks the discarded return value.
- **`delete_event` can still destroy another event's booth** when that booth sits in one of this event's rooms, through the existing `Event.rooms -> Room.booths` cascade.
  This behaves the same before and after this PR.
- `admin_edit_room` and `admin_create_booth` still store unvalidated ids, which is the root of both cross-event cases above.
  Worth closing separately; I did not touch those routes here.

## Summary by Sourcery

Prevent unauthorized event deletion and comprehensively clean up event- and room-scoped data without affecting surviving events.

Bug Fixes:
- Restrict event deletion to event owners, site administrators, and the admin token so room coordinators cannot delete an entire event.
- Ensure deleting events or rooms removes all associated memberships, transcripts, OAuth records, translation data, and other scoped rows while preserving OAuth audit history.
- Clear relay-booth references, including cross-event references, before deleting rooms or events to prevent stale pointers and deletion failures.

Enhancements:
- Add comprehensive coverage for scoped-row cleanup, OAuth audit preservation, relay relationships, cross-event references, and deletion authorization.

Tests:
- Add database and admin-panel tests covering event and room cleanup boundaries and event-owner authorization.

## Summary by CodeRabbit

* **Bug Fixes**
  * Deleting events or rooms now reliably removes associated memberships, transcripts, translations, and authorization records without leaving orphaned data.
  * Event deletion now works correctly with relay booth configurations, including cross-event links and complex booth relationships.
  * Related token references are cleared safely when events are deleted.
  * Event deletion is restricted to event owners; room coordinators can no longer delete events.
  * Audit history remains available after related authorization records are removed.

### CODE REVIEW

File: portal/database.py [hunk 1]
```diff
@@ -248,6 +291,14 @@ async def delete_room(session: AsyncSession, room_id: int) -> bool:
...
     await session.execute(sa_delete(BoothMembership).where(BoothMembership.booth_id.in_(booth_ids)))
     await session.execute(sa_delete(InviteToken).where(InviteToken.booth_id.in_(booth_ids)))
+    # Transcripts and room memberships too, or they outlive the room holding a freed
+    # room_id. delete_event only reaches these through the event's rooms, so anything
+    # stranded here can never be cleaned up later.
+    segment_scope = or_(TranscriptSegment.room_id == room_id, TranscriptSegment.booth_id.in_(booth_ids))
+    segment_ids = select(TranscriptSegment.id).where(segment_scope)
+    await session.execute(sa_delete(TranscriptTranslation).where(TranscriptTranslation.segment_id.in_(segment_ids)))
+    await session.execute(sa_delete(TranscriptSegment).where(segment_scope))
+    await session.execute(sa_delete(RoomMembership).where(RoomMembership.room_id == room_id))
```
[the remark this answers is not in the document]
Comment (arpan7sarkar) [reply portal/database.py:317]: Confirmed, and this was worse than a leftover row. Fixed in `83d189b`.

`PRAGMA foreign_key_check` after `delete_room` on `main`:

```
[('booth_translation_languages', 1, 'booths', 0)]
```

And because `delete_event` can only reach those rows through the booths, once the room was gone they were unreachable for good: deleting the event afterwards left the row still there. So it was permanent, not transient.

Verifying it also turned up a second thing in the same function, now fixed too: `delete_room` left *other* rooms relaying from the deleted room's booths pointing at booths that were about to go, and it cleared its own pointer by column rather than by relationship. `foreign_key_check` is now empty after `delete_room` on its own, not only after a following event delete.

File: portal/database.py [hunk 2]
```diff
@@ -174,6 +179,44 @@ async def delete_event(session: AsyncSession, event_id: int) -> bool:
...
+    await session.execute(sa.update(OAuthAuditLog).where(OAuthAuditLog.token_id.in_(token_ids)).values(token_id=None))
+    await session.execute(sa.update(OAuthAuditLog).where(OAuthAuditLog.event_id == event_id).values(event_id=None))
+
+    await session.execute(sa.delete(TranscriptTranslation).where(TranscriptTranslation.segment_id.in_(segment_ids)))
+    await session.execute(sa.delete(TranscriptSegment).where(segment_scope))
+    await session.execute(sa.delete(RoomMembership).where(RoomMembership.room_id.in_(room_ids)))
+    await session.execute(sa.delete(EventMembership).where(EventMembership.event_id == event_id))
+    await session.execute(sa.delete(OAuthAuthorizationCode).where(OAuthAuthorizationCode.event_id == event_id))
+    await session.execute(sa.delete(OAuthConsentGrant).where(OAuthConsentGrant.event_id == event_id))
+    await session.execute(sa.delete(OAuthToken).where(OAuthToken.event_id == event_id))
```
[the remark this answers is not in the document]
Comment (arpan7sarkar) [reply portal/database.py:224]: Applied in `83d189b`, but I want to be accurate about why, because the finding is half right.

You argued from the absent constraint, and that part holds: nothing in the schema forbids a cross-event parent, and if such a row exists this delete does strand it. I confirmed that with a hand-built row - `PRAGMA foreign_key_check` reported `[('oauth_tokens', 2, 'oauth_tokens', 0)]`.

But no current code path can create one. `parent_token_id` has exactly one writer in `portal/`, at `oauth.py:415`, and it copies `event_id=token_record.event_id` into the rotated token, so a refresh chain is always wholly inside one event and both ends are deleted together.

So this is a guard on an invariant rather than a fix for a reachable bug, which is the reason I took it: the invariant currently lives in another file, and I would rather it did not. The commit message says as much so nobody later reads it as a bug that was in the wild.

**Review by ArnavBallinCode — CHANGES_REQUESTED** [review]
# Pull Request Review: PR #463 (Event Deletion Cleanup)

I have performed a rigorous, end-to-end architectural and database-level review of PR #463. I executed the branch locally and verified the SQLite schema and constraints against `delete_event` and `delete_room`.

## Review Findings

This PR contains a critical security fix, but the core premise regarding SQLite Foreign Keys is fundamentally flawed due to recent merges on `main`.

### 1. `PRAGMA foreign_keys=ON` is Active in Production (Critical Issue)
The PR description states: *"SQLite FKs are OFF, so the ondelete= declarations never fire and only children reachable through an ORM cascade go."*

This is **factually incorrect** for production. PR #477 recently merged, which added an event listener to `portal/database.py` that executes `PRAGMA foreign_keys=ON` on every SQLite connection. 

Because FKs are ON, SQLite natively executes `ON DELETE CASCADE` and `ON DELETE SET NULL` for **all** dependent tables. I have manually verified that when `session.delete(ev)` is called:
- `Room` and `DBBooth` are cascaded.
- `TranscriptSegment`, `TranscriptTranslation`, `InviteToken`, `OAuthToken`, and all memberships are seamlessly deleted by SQLite.
- `OAuthAuditLog` automatically has its `token_id` and `event_id` fields set to `NULL`.
- Cross-event orphaned relationships (e.g., a `TranscriptSegment` tied to a booth from another event) are automatically cleaned up because the schema enforces `CASCADE` on `booth_id`.

**Result:** The 30+ lines of explicit `sa.delete()` and `sa.update()` added to `delete_event` and `delete_room` in this PR are completely redundant. They duplicate database-level constraints, introduce performance regressions via N+1 queries, and add unnecessary maintenance burden.

### 2. Test Fixture Misalignment
The reason you observed "SQLite FKs are OFF" behavior in `test_database.py` is because the `db` fixture in that file uses an isolated `create_async_engine` that **does not** include the `PRAGMA foreign_keys=ON` event listener present in `portal/database.py`. 

The correct solution is to align the test fixture with production by adding the `PRAGMA` to `test_database.py`, rather than rewriting production code to match a flawed test environment. 
*(Note: To prevent `Base.metadata.drop_all` from failing on cyclic dependencies during teardown, you can execute `PRAGMA foreign_keys=OFF` right before the `drop_all` call in the fixture).*

### 3. ORM Circular Dependency Resolution (Passed)
SQLAlchemy's Unit of Work correctly throws a `CircularDependencyError` when deleting an event if `rooms.relay_booth_id` points to `booths.id` and vice-versa, because it cannot topologically sort the deletions. Your logic to manually break the cycle by setting `room.relay_booth = None` before calling `session.delete(ev)` is correct and absolutely necessary.

### 4. Security Fix: `require_event_owner` (Passed & Critical)
Changing `require_admin` to `require_event_owner` on the `admin_delete_event` route is a vital security fix. `require_admin` dangerously admits `room_coordinator`s to any route that uses it if they belong to the event, meaning a coordinator could previously delete the entire event. Your fix and the accompanying tests (`test_delete_event_rejects_a_room_coordinator`) are excellent.

## Actionable Steps for the Author
Before this can be merged:
1. Revert all the explicit `sa.delete()` and `sa.update()` lines added to `portal/database.py`.
2. Keep the cyclic dependency breaker (`room.relay_booth = None`).
3. Keep the security fix in `portal/routers/admin.py` and its tests.
4. Update the `db` fixture in `tests/test_database.py` to enable `PRAGMA foreign_keys=ON` on connect, and disable it before `drop_all` in the teardown block.

File: portal/database.py [hunk 3]
```diff
@@ -192,12 +198,50 @@ async def delete_event(session: AsyncSession, event_id: int) -> bool:
     ev = await get_event_by_id(session, event_id)
     if ev is None:
         return False
-    # Break the circular FK cycle: rooms.relay_booth_id → booths.id ↔ booths.room_id → rooms.id
-    # assign the relationship, not the column, so an already-loaded relay_booth is
-    # cleared too and the unit of work no longer sees the Room ↔ DBBooth dependency
-    result = await session.execute(select(Room).where(Room.event_id == event_id))
-    for room in result.scalars().all():
+    # SQLite FKs are OFF, so the ondelete= declarations never fire and only children
```
Comment (Saksham-Sirohi) [inline portal/database.py:201]: This comment still says SQLite FKs are OFF, but `portal/database.py` on main attaches `PRAGMA foreign_keys=ON` on every connect (#477). Explicit sa.delete/sa.update of children duplicates ON DELETE CASCADE/SET NULL and adds maintenance cost. Please revert the redundant bulk cleanup, keep the relay_booth cycle break + flush, and enable the same PRAGMA in the `tests/test_database.py` engine fixture (disable only around drop_all teardown).

File: portal/routers/admin.py [hunk 4]
```diff
@@ -657,7 +657,7 @@ async def admin_event_api_settings_post(
     return safe_redirect(url=f"/admin/events/{event_id}/api-settings/", status_code=status.HTTP_303_SEE_OTHER)
 
 
-@router.post("/admin/events/{event_id}/delete", dependencies=[Depends(require_admin)])
+@router.post("/admin/events/{event_id}/delete", dependencies=[Depends(require_event_owner)])
```
Comment (Saksham-Sirohi) [inline portal/routers/admin.py:660]: Switching delete to `require_event_owner` is the important auth fix here — please keep it (and the coordinator/owner tests) even if the cascade cleanup is slimmed down.

**Review by Saksham-Sirohi — CHANGES_REQUESTED** [review]
Arnav CHANGES (2026-09-16) not addressed (HEAD is merge-only after review). main already enables `PRAGMA foreign_keys=ON` via #477 connect listener, but delete_event/delete_room comments and bulk sa.delete paths still assume FKs are OFF and re-implement CASCADE/SET NULL. Keep the relay cycle breaker (`room.relay_booth = None`) and the require_event_owner security fix + coordinator rejection tests — those are correct. Prefer aligning tests/test_database.py fixtures with production PRAGMA instead of duplicating DB cascades in application code.

**Review by Nirupon123 — CHANGES_REQUESTED** [review]
@arpan7sarkar please address the comments , after that we can rereview once and merge if all looks good.

**Comment by arpan7sarkar** [comment]
> @arpan7sarkar please address the comments , after that we can rereview once and merge if all looks good.

@Nirupon123  working on it
